File: cpp/test/test_binary_tree.cc
#include <bits/stdc++.h>
#include "../src/binary_tree.h"
#include "gtest.h"

TEST(BinaryTreeGenerationTest, GenerationWithPreAndInOrderTraversals)
{
    std::vector<int> preOrder = {4, 7, 9, 6, 3, 2, 5, 8, 1};
    std::vector<int> inOrder = {7, 6, 9, 3, 4, 5, 8, 2, 1};
    binary_tree::Node<int> *tree = binary_tree::Node<int>::generateWithPreAndInOrder(preOrder, inOrder);
    EXPECT_EQ(preOrder, tree->traversePreOrder());
    EXPECT_EQ(inOrder, tree->traverseInOrder());
    EXPECT_EQ(3, tree->getHeight());
    delete tree;
}

TEST(BinaryTreeGenerationTest, GenerationWithPostAndInOrderTraversals)
{
    std::vector<int> postOrder = {6, 3, 9, 7, 8, 5, 1, 2, 4};
    std::vector<int> inOrder = {7, 6, 9, 3, 4, 5, 8, 2, 1};
    binary_tree::Node<int> *tree = binary_tree::Node<int>::generateWithPostAndInOrder(postOrder, inOrder);
    EXPECT_EQ(postOrder, tree->traversePostOrder());
    EXPECT_EQ(inOrder, tree->traverseInOrder());
    EXPECT_EQ(3, tree->getHeight());
    delete tree;
}

File: src/tests/test_binary_tree.rs
#[cfg(test)]
mod tests {
    use crate::binary_tree::Node;

    #[test]
    fn test_node_set_data() {
        let mut node = Node::create(33, None, None);
        node.set_data(64);
        assert_eq!(node.get_data(), &64);
    }

    #[test]
    fn test_node_creation_by_pre_and_in_order_traversals() {
        let pre_order = [4, 7, 9, 6, 3, 2, 5, 8, 1];
        let in_order = [7, 6, 9, 3, 4, 5, 8, 2, 1];
        let tree: Node<i32> =
            Node::create_binary_tree_from_pre_and_in_order_traversals(&pre_order, &in_order);
        assert_eq!(pre_order.to_vec(), tree.traverse_pre_order());
        assert_eq!(in_order.to_vec(), tree.traverse_in_order());
    }

    #[test]
    fn test_node_creation_by_post_and_in_order_traversals() {
        let post_order = [6, 3, 9, 7, 8, 5, 1, 2, 4];
        let in_order = [7, 6, 9, 3, 4, 5, 8, 2, 1];
        let tree: Node<i32> =
            Node::create_binary_tree_from_post_and_in_order_traversals(&post_order, &in_order);
        assert_eq!(in_order.to_vec(), tree.traverse_in_order());
        assert_eq!(post_order.to_vec(), tree.traverse_post_order());
    }
}
